### PR TITLE
crimson: btree omap_manager rework with other misc fixes

### PR DESCRIPTION
**Pull request by athanatos**

### CODE REVIEW

**Comment by athanatos** [comment]
This is the first portion of the work to integrate the btree omap_manager and flat_collection_manager into seastore.

File: src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h [hunk 1]
```diff
@@ -215,31 +206,31 @@ class delta_leaf_buffer_t {
...
   }
-  void remove(std::string_view key) {
+  void remove(const std::string &key) {
     buffer.push_back(
       delta_leaf_t{
         delta_leaf_t::op_t::REMOVE,
-        key.data(),
-        ""
+        key,
+	bufferlist()
```
Comment (liu-chunmei) [inline src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h:233]: adjust indent.

File: src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h [hunk 2]
```diff
@@ -215,31 +206,31 @@ class delta_leaf_buffer_t {
...
-    std::string_view val) {
+    const std::string &key,
+    const ceph::bufferlist &val) {
     buffer.push_back(
       delta_leaf_t{
         delta_leaf_t::op_t::UPDATE,
-        key.data(),
-        val.data()
+	key,
+	val
```
Comment (liu-chunmei) [inline src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h:225]: please check indent.

File: src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h [hunk 3]
```diff
@@ -146,35 +147,31 @@ class delta_inner_buffer_t {
...
   }
-  void remove(std::string_view val) {
+  void remove(const std::string &key) {
     buffer.push_back(
       delta_inner_t{
         delta_inner_t::op_t::REMOVE,
-        omap_inner_key_le_t(),
-        val.data()
+	key,
+	L_ADDR_NULL
```
Comment (liu-chunmei) [inline src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h:174]: I am not sure if the indent is you want, please check.

File: src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h [hunk 4]
```diff
@@ -146,35 +147,31 @@ class delta_inner_buffer_t {
...
-    k = key;
+    const std::string &key,
+    laddr_t addr) {
     buffer.push_back(
       delta_inner_t{
         delta_inner_t::op_t::UPDATE,
-        k,
-        val.data()
+        key,
+	addr
```
Comment (liu-chunmei) [inline src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h:166]: please check indent.

File: src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h [hunk 5]
```diff
@@ -1341,9 +1241,9 @@ class StringKVLeafNodeLayout {
     auto iter = iter_begin();
     auto iter2 = rhs.iter_begin();
     while (iter != iter_end()) {
-      if(iter->get_node_key() != iter2->get_node_key() ||
-	       iter->get_node_val() != iter2->get_node_val() ||
-         iter->get_string_val() != iter2->get_string_val()){
+      if(iter->get_key() != iter2->get_key() ||
+	       iter->get_key() != iter2->get_key() ||
+         iter->get_val() != iter2->get_val()){
 	      return false;
```
[the remark this answers is not in the document]
Comment (athanatos) [reply src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h:1247]: Good catch.  Going to remove the second get_key() variant.  The previous version also checked the node_key, but that isn't really necessary here.

File: src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h [hunk 6]
```diff
@@ -1341,10 +1241,9 @@ class StringKVLeafNodeLayout {
     auto iter = iter_begin();
     auto iter2 = rhs.iter_begin();
     while (iter != iter_end()) {
-      if(iter->get_node_key() != iter2->get_node_key() ||
-	       iter->get_node_val() != iter2->get_node_val() ||
-         iter->get_string_val() != iter2->get_string_val()){
-	      return false;
+      if(iter->get_key() != iter2->get_key() ||
+         iter->get_val() != iter2->get_val()) {
```
Comment (liu-chunmei) [inline src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h:1245]: get_val() return bufferlist,  I can't find bufferlist operator!= definition, should it be get_string_val() here?

Comment (athanatos) [reply src/crimson/os/seastore/omap_manager/btree/string_kv_node_layout.h:1245]: get_val() returns the mapped value for the key at iter, which happens to be a bufferlist.  operator!= is defined for bufferlists in include/buffer.h.

**Review by liu-chunmei — APPROVED** [review]

**Review by cyx1231st — APPROVED** [review]

**Review by tchaikov — APPROVED** [review]

**Pull request merged**